datasets/pcfg/data_generation.py

```diff
@@ -0,0 +1,124 @@
+""" Generating training and test files for iterative decoding on PCFG.
+"""
+
+import re
+import numpy as np
+
+# Specify local path to PCFG data in
+# https://github.com/i-machine-think/am-i-compositional/tree/master/data/pcfgset
+# Choose between iid (pcfgset), productivity and systematicity splits
+data_path = "/content/drive/MyDrive/preprocess_pcfg/pcfgset/data/"
+
+# Function that generates a list of intermediate tasks (including original source 
+# and original target) for each source line and target line pair
+def generate_intermediate_tasks(source_line, target_line):
+
+  unary_keys = ['copy', 'reverse', 'shift', 'swap_first_last', 'repeat', 'echo']
+  binary_keys =['append', 'prepend', 'remove_first', 'remove_second']
+  key_lengths = {k: len(k) for k in unary_keys + binary_keys}
+
+  # Listing all the operations in a sentence (ordered from right to left)
+  op_idx = []
+  operations = []
+  for k in unary_keys + binary_keys:
+    this_op_idx = [m.start() for m in re.finditer(k, source_line)]
+    for i in this_op_idx:
+      op_idx.append(i)
+      operations.append(k)
+  op_idx = np.array(op_idx)
+  sort_idx = np.argsort(-op_idx)
+  op_idx = op_idx[sort_idx]
+  operation_list = [operations[i] for i in sort_idx.tolist()]
+
+  # Populating list of intermediate tasks
+  intermediate_tasks = [source_line]
+  aux_line = source_line
+  for i, op in zip(op_idx, operation_list):
+    this_int_task = aux_line[0:i]
+    right_idx = i + key_lengths[op] + 1
+    aux_op = aux_line[right_idx:]
+    # Unary keys
+    if op in unary_keys:
+      left_idx = aux_op.find(' , ')
+      if left_idx != -1:
+        aux_op = aux_op[:left_idx]
+      aux_list = aux_op.split(' ')
+      # Copy or repeat
+      if op == 'copy' or op == 'repeat':
+        this_int_task = this_int_task + aux_op
+        if op == 'repeat':
+          this_int_task = this_int_task + ' ' + aux_op
+        if left_idx != -1:  
+          this_int_task += ' ' + aux_line[right_idx+left_idx+1:]
+      # Other unary operations
+      else:
+        if op == 'reverse':
+          aux_list = aux_list[::-1]
+        elif op == 'shift':
+          shifted_elt = aux_list.pop(0)
+          aux_list.append(shifted_elt)
+        elif op == 'swap_first_last':
+          aux_list[0], aux_list[-1] = aux_list[-1], aux_list[0]
+        else: # op == 'echo'
+          aux_list.append(aux_list[-1])
+        for elt in aux_list:
```

Comment:
I don't think I understand very well what these lines do, a comment will help! Also, what does "elt" stand for?